`AHashMap::default()` gives a type inferrence error

(again, please ignore FxHashMap - it's really an AHashMap)
```
error[E0282]: type annotations needed for `FxHashMap<usize, snippet::Style, S>`
    --> compiler/rustc_errors/src/emitter.rs:1356:38
     |
1356 |                 let mut multilines = FxHashMap::default();
     |                     --------------   ^^^^^^^^^^^^^^^^^^ cannot infer type for type parameter `S` declared on the struct `AHashMap`
     |                     |
     |                     consider giving `multilines` the explicit type `FxHashMap<usize, snippet::Style, S>`, where the type parameter `S` is specified
```
The way `FxHashMap` deals with this is by making `FxHashMap` a straight type alias for using `BuildHasherDefault`: https://docs.rs/rustc-hash/1.1.0/rustc_hash/type.FxHashMap.html. Maybe `AHashMap` could do something similar?

`new()` has the same issue.
```
error[E0283]: type annotations needed for `FxHashMap<usize, snippet::Style, S>`
    --> compiler/rustc_errors/src/emitter.rs:1357:38
     |
1357 |                 let mut multilines = FxHashMap::new();
     |                     --------------   ^^^^^^^^^^^^^^ cannot infer type for type parameter `S`
     |                     |
     |                     consider giving `multilines` the explicit type `FxHashMap<usize, snippet::Style, S>`, where the type parameter `S` is specified
     |
     = note: cannot satisfy `_: BuildHasher`
     = note: required by `FxHashMap::<K, V, S>::new`
```

@tkaitchuck indeed it does! Now it runs into other errors, but those I need to fix myself I think.

```
error[E0277]: the trait bound `FxHashMap<DepNode<K>, SerializedDepNodeIndex>: Encodable<_>` is not satisfied
  --> compiler/rustc_query_system/src/dep_graph/prev.rs:6:17
   |
6  |   #[derive(Debug, Encodable, Decodable)]
   |                   ^^^^^^^^^
   |                   |
   |                   the trait `Encodable<_>` is not implemented for `FxHashMap<DepNode<K>, SerializedDepNodeIndex>`
   |                   in this macro invocation
   | 
  ::: /home/joshua/.local/lib/cargo/registry/src/github.com-1ecc6299db9ec823/synstructure-0.12.4/src/macros.rs:94:9
   |
94 | /         pub fn $derives(
95 | |             i: $crate::macros::TokenStream
96 | |         ) -> $crate::macros::TokenStream {
   | |________________________________________- in this expansion of `#[derive(Encodable)]`
   |
   = note: required by `rustc_serialize::Encodable::encode`
```

This error I don't understand ... it looks like `insert()` takes a different mutability than before? But `FxHashMap` also takes `&mut self` for insert(), so I don't know what could have changed.

```
error[E0502]: cannot borrow `*self` as immutable because it is also borrowed as mutable
  --> compiler/rustc_resolve/src/def_collector.rs:53:92
   |
53 |             let old_index = self.resolver.placeholder_field_indices.insert(field.id, index(self));
   |                             --------------------------------------- ------                 ^^^^ immutable borrow occurs here
   |                             |                                       |
   |                             |                                       mutable borrow later used by call
   |                             mutable borrow occurs here

error: aborting due to previous error
```

I'm also seeing errors like this, which might be related to the `K: Borrow<Q>` bound on `get()`? Not sure why that would be an error, though.
```
error[E0308]: mismatched types
   --> compiler/rustc_mir/src/interpret/memory.rs:627:30
    |
627 |                         .get(&id)
    |                              ^^^
    |                              |
    |                              expected struct `rustc_middle::mir::interpret::AllocId`, found `&rustc_middle::mir::interpret::AllocId`
    |                              help: consider removing the borrow: `id`
```

This also happens with the other codebase I tried to switch to ahash! (well, from hashbrown::HashMap to ahash::AHashMap.) https://github.com/launchbadge/sqlx/pull/737

```
error[E0502]: cannot borrow `n` as immutable because it is also borrowed as mutable
   --> sqlx-core/src/sqlite/connection/explain.rs:139:34
    |
139 |                     n.insert(p3, n.get(&p2).copied().unwrap_or(true));
    |                     - ------     ^ immutable borrow occurs here
    |                     | |
    |                     | mutable borrow later used by call
    |                     mutable borrow occurs here
```

> error[E0502]: cannot borrow `n` as immutable because it is also borrowed as mutable
   --> sqlx-core/src/sqlite/connection/explain.rs:139:34

I wonder if it's because of the deref trait, and if it would work better if the class directly implemented insert and get. I'll try it out.

Works great :) There was another error I don't think you can solve without getting rid of deref altogether:
```
error: function cannot return without recursing
   --> compiler/rustc_mir/src/const_eval/machine.rs:120:5
    |
120 | /     fn contains_key<Q: ?Sized + Hash + Eq>(&mut self, k: &Q) -> bool
121 | |     where
122 | |         K: Borrow<Q>,
123 | |     {
124 | |         FxHashMap::contains_key(self, k)
    | |         -------------------------------- recursive call site
125 | |     }
    | |_____^ cannot return without recursing
```
What happened is this is in a trait implementation; previously this would call the inherent method, but now it calls the method from the trait. Fortunately rustc has good warnings and this is the fix:
```diff
diff --git a/compiler/rustc_mir/src/const_eval/machine.rs b/compiler/rustc_mir/src/const_eval/machine.rs
index 73ca7e0d471..b7a142ce330 100644
--- a/compiler/rustc_mir/src/const_eval/machine.rs
+++ b/compiler/rustc_mir/src/const_eval/machine.rs
@@ -121,7 +121,7 @@ impl<K: Hash + Eq, V> interpret::AllocMap<K, V> for FxHashMap<K, V> {
     where
         K: Borrow<Q>,
     {
-        FxHashMap::contains_key(self, k)
+        (**self).contains_key(k)
     }
 
     #[inline(always)]
```

This error looks a little odd, though:
```
error[E0308]: mismatched types
   --> src/tools/clippy/clippy_lints/src/copies.rs:393:58
    |
393 |         FxHashMap::with_capacity_and_hasher(exprs.len(), BuildHasherDefault::default());
    |                                                          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ expected struct `ahash::random_state::RandomState`, found struct `BuildHasherDefault`
    |
    = note: expected struct `ahash::random_state::RandomState`
               found struct `BuildHasherDefault<_>`
```

I think you might have changed _too_ many things to using RandomState? I wouldn't expect `with_hasher` to require a specific type.

--- a/src/hash_map.rs
+++ b/src/hash_map.rs
@@ -13,21 +13,19 @@ use crate::RandomState;
 #[derive(Clone)]
 pub struct AHashMap<K, V, S = crate::RandomState>(HashMap<K, V, S>);
 
-impl<K, V> From<HashMap<K,V,crate::RandomState>> for AHashMap<K, V> {
-    fn from(item: HashMap<K,V,crate::RandomState>) -> Self {
+impl<K, V> From<HashMap<K, V, crate::RandomState>> for AHashMap<K, V> {
+    fn from(item: HashMap<K, V, crate::RandomState>) -> Self {
         AHashMap(item)
     }
 }
 
-impl<K, V> Into<HashMap<K,V,crate::RandomState>> for AHashMap<K, V> {
-    fn into(self) -> HashMap<K,V,crate::RandomState> {
+impl<K, V> Into<HashMap<K, V, crate::RandomState>> for AHashMap<K, V> {
+    fn into(self) -> HashMap<K, V, crate::RandomState> {
         self.0
     }
 }
 
-impl<K, V> AHashMap<K, V, RandomState>
-where
-{
+impl<K, V> AHashMap<K, V, RandomState> {
     pub fn new() -> Self {
         AHashMap(HashMap::with_hasher(RandomState::default()))
     }
@@ -42,6 +40,143 @@ where
 
     pub fn with_capacity_and_hasher(capacity: usize, hash_builder: RandomState) -> Self {
         AHashMap(HashMap::with_capacity_and_hasher(capacity, hash_builder))
+    }
+}
+
+impl<K, V> AHashMap<K, V, RandomState>
+where
+    K: Hash + Eq,
+{
+    /// Returns a reference to the value corresponding to the key.
+    ///
+    /// The key may be any borrowed form of the map's key type, but
+    /// [`Hash`] and [`Eq`] on the borrowed form *must* match those for
+    /// the key type.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use std::collections::HashMap;
+    ///
+    /// let mut map = HashMap::new();
+    /// map.insert(1, "a");
+    /// assert_eq!(map.get(&1), Some(&"a"));
+    /// assert_eq!(map.get(&2), None);
+    /// ```
+    #[inline]
+    pub fn get<Q: ?Sized>(&self, k: &Q) -> Option<&V>
+    where
+        K: Borrow<Q>,
+        Q: Hash + Eq,
+    {
+        self.0.get(k)
+    }
+
+    /// Returns the key-value pair corresponding to the supplied key.
+    ///
+    /// The supplied key may be any borrowed form of the map's key type, but
+    /// [`Hash`] and [`Eq`] on the borrowed form *must* match those for
+    /// the key type.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use std::collections::HashMap;
+    ///
+    /// let mut map = HashMap::new();
+    /// map.insert(1, "a");
+    /// assert_eq!(map.get_key_value(&1), Some((&1, &"a")));
+    /// assert_eq!(map.get_key_value(&2), None);
+    /// ```
+    #[inline]
+    pub fn get_key_value<Q: ?Sized>(&self, k: &Q) -> Option<(&K, &V)>
+    where
+        K: Borrow<Q>,
+        Q: Hash + Eq,
+    {
+        self.0.get_key_value(k)
+    }
+
+    /// Returns a mutable reference to the value corresponding to the key.
+    ///
+    /// The key may be any borrowed form of the map's key type, but
+    /// [`Hash`] and [`Eq`] on the borrowed form *must* match those for
+    /// the key type.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use std::collections::HashMap;
+    ///
+    /// let mut map = HashMap::new();
+    /// map.insert(1, "a");
+    /// if let Some(x) = map.get_mut(&1) {
+    ///     *x = "b";
+    /// }
+    /// assert_eq!(map[&1], "b");
+    /// ```
+    #[inline]
+    pub fn get_mut<Q: ?Sized>(&mut self, k: &Q) -> Option<&mut V>
+    where
+        K: Borrow<Q>,
+        Q: Hash + Eq,
+    {
+        self.0.get_mut(k)
+    }
+
+    /// Inserts a key-value pair into the map.
+    ///
+    /// If the map did not have this key present, [`None`] is returned.
+    ///
+    /// If the map did have this key present, the value is updated, and the old
+    /// value is returned. The key is not updated, though; this matters for
+    /// types that can be `==` without being identical. See the [module-level
+    /// documentation] for more.
+    ///
+    /// [module-level documentation]: crate::collections#insert-and-complex-keys
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use std::collections::HashMap;
+    ///
+    /// let mut map = HashMap::new();
+    /// assert_eq!(map.insert(37, "a"), None);
+    /// assert_eq!(map.is_empty(), false);
+    ///
+    /// map.insert(37, "b");
+    /// assert_eq!(map.insert(37, "c"), Some("b"));
+    /// assert_eq!(map[&37], "c");
+    /// ```
+    #[inline]
+    pub fn insert(&mut self, k: K, v: V) -> Option<V> {
+        self.0.insert(k, v)
+    }
+
+    /// Removes a key from the map, returning the value at the key if the key
+    /// was previously in the map.
+    ///
+    /// The key may be any borrowed form of the map's key type, but
+    /// [`Hash`] and [`Eq`] on the borrowed form *must* match those for
+    /// the key type.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use std::collections::HashMap;
+    ///
+    /// let mut map = HashMap::new();
+    /// map.insert(1, "a");
+    /// assert_eq!(map.remove(&1), Some("a"));
+    /// assert_eq!(map.remove(&1), None);
+    /// ```
+    #[inline]
+    pub fn remove<Q: ?Sized>(&mut self, k: &Q) -> Option<V>
+    where
+        K: Borrow<Q>,
+        Q: Hash + Eq,
+    {
+        self.0.remove(k)
     }
 }
 
@@ -175,5 +310,16 @@ impl<K, V> Default for AHashMap<K, V, RandomState> {
     #[inline]
     fn default() -> AHashMap<K, V, RandomState> {
         AHashMap::new()
+    }
+}
+
+#[cfg(test)]
+mod test {
+    use super::*;
+    #[test]
+    fn test_borrow() {
+        let mut map: AHashMap<String, String> = AHashMap::new();
+        map.insert("foo".to_string(), "Bar".to_string());
+        map.insert("Bar".to_string(), map.get("foo").unwrap().to_owned());
     }
 }

--- a/src/hash_map.rs
+++ b/src/hash_map.rs
@@ -43,9 +43,10 @@ impl<K, V> AHashMap<K, V, RandomState> {
     }
 }
 
-impl<K, V> AHashMap<K, V, RandomState>
+impl<K, V, S> AHashMap<K, V, S>
 where
     K: Hash + Eq,
+    S: BuildHasher
 {
     /// Returns a reference to the value corresponding to the key.
     ///

--- a/src/hash_map.rs
+++ b/src/hash_map.rs
@@ -1,12 +1,12 @@
 use std::borrow::Borrow;
 use std::collections::{hash_map, HashMap};
 use std::fmt::{self, Debug};
-use std::hash::{BuildHasher, Hash};
+use std::hash::{BuildHasher, Hash, BuildHasherDefault};
 use std::iter::FromIterator;
 use std::ops::{Deref, DerefMut, Index};
 use std::panic::UnwindSafe;
 
-use crate::RandomState;
+use crate::{RandomState, AHasher};
 
 /// A [`HashMap`](std::collections::HashMap) using [`RandomState`](crate::RandomState) to hash the items.
 /// (Requires the `std` feature to be enabled.)
@@ -39,6 +39,17 @@ impl<K, V> AHashMap<K, V, RandomState> {
     }
 
     pub fn with_capacity_and_hasher(capacity: usize, hash_builder: RandomState) -> Self {
+        AHashMap(HashMap::with_capacity_and_hasher(capacity, hash_builder))
+    }
+}
+
+impl<K, V> AHashMap<K, V, BuildHasherDefault<AHasher>> {
+
+    pub fn with_hasher(hash_builder: BuildHasherDefault<AHasher>) -> Self {
+        AHashMap(HashMap::with_hasher(hash_builder))
+    }
+
+    pub fn with_capacity_and_hasher(capacity: usize, hash_builder: BuildHasherDefault<AHasher>) -> Self {
         AHashMap(HashMap::with_capacity_and_hasher(capacity, hash_builder))
     }
 }

--- a/src/hash_map.rs
+++ b/src/hash_map.rs
@@ -6,6 +6,8 @@ use std::iter::FromIterator;
 use std::ops::{Deref, DerefMut, Index};
 use std::panic::UnwindSafe;
 
+use crate::RandomState;
+
 /// A [`HashMap`](std::collections::HashMap) using [`RandomState`](crate::RandomState) to hash the items.
 /// (Requires the `std` feature to be enabled.)
 #[derive(Clone)]
@@ -23,28 +25,22 @@ impl<K, V> Into<HashMap<K,V,crate::RandomState>> for AHashMap<K, V> {
     }
 }
 
-impl<K, V, S> AHashMap<K, V, S>
+impl<K, V> AHashMap<K, V, RandomState>
 where
-    S: BuildHasher + Default,
 {
     pub fn new() -> Self {
-        AHashMap(HashMap::with_hasher(S::default()))
+        AHashMap(HashMap::with_hasher(RandomState::default()))
     }
 
     pub fn with_capacity(capacity: usize) -> Self {
-        AHashMap(HashMap::with_capacity_and_hasher(capacity, S::default()))
+        AHashMap(HashMap::with_capacity_and_hasher(capacity, RandomState::default()))
     }
-}
 
-impl<K, V, S> AHashMap<K, V, S>
-where
-    S: BuildHasher,
-{
-    pub fn with_hasher(hash_builder: S) -> Self {
+    pub fn with_hasher(hash_builder: RandomState) -> Self {
         AHashMap(HashMap::with_hasher(hash_builder))
     }
 
-    pub fn with_capacity_and_hasher(capacity: usize, hash_builder: S) -> Self {
+    pub fn with_capacity_and_hasher(capacity: usize, hash_builder: RandomState) -> Self {
         AHashMap(HashMap::with_capacity_and_hasher(capacity, hash_builder))
     }
 }
@@ -175,12 +171,9 @@ where
     }
 }
 
-impl<K, V, S> Default for AHashMap<K, V, S>
-where
-    S: BuildHasher + Default,
-{
+impl<K, V> Default for AHashMap<K, V, RandomState> {
     #[inline]
-    fn default() -> AHashMap<K, V, S> {
-        AHashMap::with_hasher(Default::default())
+    fn default() -> AHashMap<K, V, RandomState> {
+        AHashMap::new()
     }
 }

--- a/src/hash_map.rs
+++ b/src/hash_map.rs
@@ -1,12 +1,12 @@
 use std::borrow::Borrow;
 use std::collections::{hash_map, HashMap};
 use std::fmt::{self, Debug};
-use std::hash::{BuildHasher, Hash, BuildHasherDefault};
+use std::hash::{BuildHasher, Hash};
 use std::iter::FromIterator;
 use std::ops::{Deref, DerefMut, Index};
 use std::panic::UnwindSafe;
 
-use crate::{RandomState, AHasher};
+use crate::{RandomState};
 
 /// A [`HashMap`](std::collections::HashMap) using [`RandomState`](crate::RandomState) to hash the items.
 /// (Requires the `std` feature to be enabled.)
@@ -33,23 +33,15 @@ impl<K, V> AHashMap<K, V, RandomState> {
     pub fn with_capacity(capacity: usize) -> Self {
         AHashMap(HashMap::with_capacity_and_hasher(capacity, RandomState::default()))
     }
-
-    pub fn with_hasher(hash_builder: RandomState) -> Self {
-        AHashMap(HashMap::with_hasher(hash_builder))
-    }
-
-    pub fn with_capacity_and_hasher(capacity: usize, hash_builder: RandomState) -> Self {
-        AHashMap(HashMap::with_capacity_and_hasher(capacity, hash_builder))
-    }
 }
 
-impl<K, V> AHashMap<K, V, BuildHasherDefault<AHasher>> {
+impl<K, V, S> AHashMap<K, V, S> where S: BuildHasher {
 
-    pub fn with_hasher(hash_builder: BuildHasherDefault<AHasher>) -> Self {
+    pub fn with_hasher(hash_builder: S) -> Self {
         AHashMap(HashMap::with_hasher(hash_builder))
     }
 
-    pub fn with_capacity_and_hasher(capacity: usize, hash_builder: BuildHasherDefault<AHasher>) -> Self {
+    pub fn with_capacity_and_hasher(capacity: usize, hash_builder: S) -> Self {
         AHashMap(HashMap::with_capacity_and_hasher(capacity, hash_builder))
     }
 }

--- a/src/hash_set.rs
+++ b/src/hash_set.rs
@@ -1,7 +1,7 @@
-use crate::{RandomState, AHasher};
+use crate::{RandomState};
 use std::collections::{hash_set, HashSet};
 use std::fmt::{self, Debug};
-use std::hash::{BuildHasher, Hash, BuildHasherDefault};
+use std::hash::{BuildHasher, Hash};
 use std::iter::FromIterator;
 use std::ops::{BitAnd, BitOr, BitXor, Deref, DerefMut, Sub};
 
@@ -30,23 +30,15 @@ impl<T> AHashSet<T, RandomState> {
     pub fn with_capacity(capacity: usize) -> Self {
         AHashSet(HashSet::with_capacity_and_hasher(capacity, RandomState::default()))
     }
-
-    pub fn with_hasher(hash_builder: RandomState) -> Self {
-        AHashSet(HashSet::with_hasher(hash_builder))
-    }
-
-    pub fn with_capacity_and_hasher(capacity: usize, hash_builder: RandomState) -> Self {
-        AHashSet(HashSet::with_capacity_and_hasher(capacity, hash_builder))
-    }
 }
 
-impl<T> AHashSet<T, BuildHasherDefault<AHasher>> {
+impl<T, S> AHashSet<T, S> where S: BuildHasher {
 
-    pub fn with_hasher(hash_builder: BuildHasherDefault<AHasher>) -> Self {
+    pub fn with_hasher(hash_builder: S) -> Self {
         AHashSet(HashSet::with_hasher(hash_builder))
     }
 
-    pub fn with_capacity_and_hasher(capacity: usize, hash_builder: BuildHasherDefault<AHasher>) -> Self {
+    pub fn with_capacity_and_hasher(capacity: usize, hash_builder: S) -> Self {
         AHashSet(HashSet::with_capacity_and_hasher(capacity, hash_builder))
     }
 }

--- a/src/hash_set.rs
+++ b/src/hash_set.rs
@@ -1,23 +1,23 @@
+use crate::RandomState;
 use std::collections::{hash_set, HashSet};
 use std::fmt::{self, Debug};
 use std::hash::{BuildHasher, Hash};
 use std::iter::FromIterator;
 use std::ops::{BitAnd, BitOr, BitXor, Deref, DerefMut, Sub};
-use crate::RandomState;
 
 /// A [`HashSet`](std::collections::HashSet) using [`RandomState`](crate::RandomState) to hash the items.
 /// (Requires the `std` feature to be enabled.)
 #[derive(Clone)]
 pub struct AHashSet<T, S = crate::RandomState>(HashSet<T, S>);
 
-impl<T> From<HashSet<T,crate::RandomState>> for AHashSet<T> {
-    fn from(item: HashSet<T,crate::RandomState>) -> Self {
+impl<T> From<HashSet<T, crate::RandomState>> for AHashSet<T> {
+    fn from(item: HashSet<T, crate::RandomState>) -> Self {
         AHashSet(item)
     }
 }
 
-impl<T> Into<HashSet<T,crate::RandomState>> for AHashSet<T> {
-    fn into(self) -> HashSet<T,crate::RandomState> {
+impl<T> Into<HashSet<T, crate::RandomState>> for AHashSet<T> {
+    fn into(self) -> HashSet<T, crate::RandomState> {
         self.0
     }
 }

--- a/src/hash_set.rs
+++ b/src/hash_set.rs
@@ -1,7 +1,7 @@
-use crate::RandomState;
+use crate::{RandomState, AHasher};
 use std::collections::{hash_set, HashSet};
 use std::fmt::{self, Debug};
-use std::hash::{BuildHasher, Hash};
+use std::hash::{BuildHasher, Hash, BuildHasherDefault};
 use std::iter::FromIterator;
 use std::ops::{BitAnd, BitOr, BitXor, Deref, DerefMut, Sub};
 
@@ -36,6 +36,17 @@ impl<T> AHashSet<T, RandomState> {
     }
 
     pub fn with_capacity_and_hasher(capacity: usize, hash_builder: RandomState) -> Self {
+        AHashSet(HashSet::with_capacity_and_hasher(capacity, hash_builder))
+    }
+}
+
+impl<T> AHashSet<T, BuildHasherDefault<AHasher>> {
+
+    pub fn with_hasher(hash_builder: BuildHasherDefault<AHasher>) -> Self {
+        AHashSet(HashSet::with_hasher(hash_builder))
+    }
+
+    pub fn with_capacity_and_hasher(capacity: usize, hash_builder: BuildHasherDefault<AHasher>) -> Self {
         AHashSet(HashSet::with_capacity_and_hasher(capacity, hash_builder))
     }
 }

--- a/src/hash_set.rs
+++ b/src/hash_set.rs
@@ -3,6 +3,7 @@ use std::fmt::{self, Debug};
 use std::hash::{BuildHasher, Hash};
 use std::iter::FromIterator;
 use std::ops::{BitAnd, BitOr, BitXor, Deref, DerefMut, Sub};
+use crate::RandomState;
 
 /// A [`HashSet`](std::collections::HashSet) using [`RandomState`](crate::RandomState) to hash the items.
 /// (Requires the `std` feature to be enabled.)
@@ -21,28 +22,20 @@ impl<T> Into<HashSet<T,crate::RandomState>> for AHashSet<T> {
     }
 }
 
-impl<T, S> AHashSet<T, S>
-where
-    S: BuildHasher + Default,
-{
+impl<T> AHashSet<T, RandomState> {
     pub fn new() -> Self {
-        AHashSet(HashSet::with_hasher(S::default()))
+        AHashSet(HashSet::with_hasher(RandomState::default()))
     }
 
     pub fn with_capacity(capacity: usize) -> Self {
-        AHashSet(HashSet::with_capacity_and_hasher(capacity, S::default()))
+        AHashSet(HashSet::with_capacity_and_hasher(capacity, RandomState::default()))
     }
-}
 
-impl<T, S> AHashSet<T, S>
-where
-    S: BuildHasher,
-{
-    pub fn with_hasher(hash_builder: S) -> Self {
+    pub fn with_hasher(hash_builder: RandomState) -> Self {
         AHashSet(HashSet::with_hasher(hash_builder))
     }
 
-    pub fn with_capacity_and_hasher(capacity: usize, hash_builder: S) -> Self {
+    pub fn with_capacity_and_hasher(capacity: usize, hash_builder: RandomState) -> Self {
         AHashSet(HashSet::with_capacity_and_hasher(capacity, hash_builder))
     }
 }
@@ -264,13 +257,10 @@ where
     }
 }
 
-impl<T, S> Default for AHashSet<T, S>
-where
-    S: BuildHasher + Default,
-{
+impl<T> Default for AHashSet<T, RandomState> {
     /// Creates an empty `AHashSet<T, S>` with the `Default` value for the hasher.
     #[inline]
-    fn default() -> AHashSet<T, S> {
+    fn default() -> AHashSet<T, RandomState> {
         AHashSet(HashSet::default())
     }
 }

